Update pytest and sybil

Marenz opened the pull request:
fixes #219

Marenz committed:
Update pytest and sybil

* pytest 8.0.0
* sybil 6.0.3

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz commented:
Skipping release notes for this dependency-only update, let me know if you want notes.

Marenz commented:
Not sure what the CIs problem is really

llucax commented:
I should have mentioned this, but I also had some work done on it, I think I also bumped into tests issues and got distracted with something else and dropped it.

Reading the tests is hard because the tests do generate a whole project for each type of repo, so you have to dig in all the ton of output to look for what it failed.

Besides some weirdness when running `git` ([1](https://github.com/frequenz-floss/frequenz-repo-config-python/actions/runs/7784779454/job/21225967611?pr=221#step:8:677), [2](https://github.com/frequenz-floss/frequenz-repo-config-python/actions/runs/7784779454/job/21225967611?pr=221#step:8:699)) that seemed to be non-fatal (despite some message saying fatal), the [real problem](https://github.com/frequenz-floss/frequenz-repo-config-python/actions/runs/7784779454/job/21225967611?pr=221#step:8:894) seems to be:

```
INFO: pip is looking at multiple versions of frequenz-repo-config[extra-lint-examples] to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install frequenz-actor-test, frequenz-actor-test[dev-flake8,dev-formatting,dev-mkdocs,dev-mypy,dev-noxfile,dev-pylint,dev-pytest]==0.0.post0+d20240205, frequenz-actor-test[dev-mkdocs,dev-noxfile,dev-pytest]==0.0.post0+d20240205 and frequenz-repo-config[extra-lint-examples]==0.0.post1+g3a6a060 because these package versions have conflicting dependencies.

The conflict is caused by:
    frequenz-actor-test[dev-flake8,dev-formatting,dev-mkdocs,dev-mypy,dev-noxfile,dev-pylint,dev-pytest] 0.0.post0+d20240205 depends on pytest==7.4.2; extra == "dev-pytest"
    pytest-asyncio 0.21.1 depends on pytest>=7.0.0
    pytest-mock 3.11.1 depends on pytest>=5.0
    frequenz-actor-test[dev-mkdocs,dev-noxfile,dev-pytest] 0.0.post0+d20240205 depends on pytest==7.4.2; extra == "dev-pytest"
    frequenz-repo-config[extra-lint-examples] 0.0.post1+g3a6a060 depends on pytest<9 and >=8; extra == "extra-lint-examples"

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

Marenz commented:
I can't run it successfully locally, but the CI is happy :laughing:

llucax commented:
Seems to work locally for me using `nox` (so regenerating all nox venvs). It is still running the tests locally, but the dependency issue is not there, it should have happen before running the tests).

Maybe you are using `nox -R` or just running in your own venv which has the dependencies not fully updated?

llucax committed:
Widen dependencies instead of bumping them

The `pytest` and `sybil` dependencies can be purely widen, supporting
the same oldest versions as before, as we are not affected by the
breaking changes.

This allows us to keep backwards compatibility for downstream project.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented on pyproject.toml:
```diff
@@ -63,8 +63,8 @@ lib = []
   "pylint >= 2.17.3, < 4",
-  "pytest >= 7.3.0, < 8",
-  "sybil >= 5.0.3, < 6",
+  "pytest >= 7.4.2, < 9",
+  "sybil >= 6.0.3, < 7",
```
Ideally we should only widen these to avoid making this a breaking change and still support pytest 7.x (if we are not affected by the breaking changes in 8):

```suggestion
  "pytest >= 7.3.0, < 9",
  "sybil >= 5.0.3, < 7",
```

I did some local testing with this and it seems to work.

llucax reviewed (commented):
Without the suggested change, this would also needs a note in the release notes and upgrade instructions as bumping the dependency is a breaking change (downstream projects will need to upgrade too). But again, locally it works with only widening the dependencies for `extra-lint-examples`.

Sybil 6 has type information, so we can also remove the `mypy` exception for it.

Here is a patch of local changes I made that still pass all tests:

```diff
diff --git a/cookiecutter/{{cookiecutter.github_repo_name}}/pyproject.toml b/cookiecutter/{{cookiecutter.github_repo_name}}/pyproject.toml
index c603793..d63cd7b 100644
--- a/cookiecutter/{{cookiecutter.github_repo_name}}/pyproject.toml
+++ b/cookiecutter/{{cookiecutter.github_repo_name}}/pyproject.toml
@@ -193,7 +193,7 @@ packages = ["{{cookiecutter.python_package}}"]
 strict = true
 
 [[tool.mypy.overrides]]
-module = ["mkdocs_macros.*", "sybil", "sybil.*"]
+module = ["mkdocs_macros.*"]
 ignore_missing_imports = true
 {%- if cookiecutter.type == "api" %}
 
diff --git a/pyproject.toml b/pyproject.toml
index b396bf2..022ef0d 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -63,8 +63,8 @@ lib = []
 model = []
 extra-lint-examples = [
   "pylint >= 2.17.3, < 4",
-  "pytest >= 7.4.2, < 9",
-  "sybil >= 6.0.3, < 7",
+  "pytest >= 7.3.0, < 9",
+  "sybil >= 5.0.3, < 7",
 ]
 dev-flake8 = [
   "flake8 == 6.1.0",
@@ -188,8 +188,6 @@ module = [
   "github_action_utils",
   "mkdocs_macros.*",
   "semver.version",
-  "sybil",
-  "sybil.*",
 ]
 ignore_missing_imports = true
 
diff --git a/tests_golden/integration/test_cookiecutter_generation/actor/frequenz-actor-test/pyproject.toml b/tests_golden/integration/test_cookiecutter_generation/actor/frequenz-actor-test/pyproject.toml
index 0cbe3ee..f12ef54 100644
--- a/tests_golden/integration/test_cookiecutter_generation/actor/frequenz-actor-test/pyproject.toml
+++ b/tests_golden/integration/test_cookiecutter_generation/actor/frequenz-actor-test/pyproject.toml
@@ -163,7 +163,7 @@ packages = ["frequenz.actor.test"]
 strict = true
 
 [[tool.mypy.overrides]]
-module = ["mkdocs_macros.*", "sybil", "sybil.*"]
+module = ["mkdocs_macros.*"]
 ignore_missing_imports = true
 
 [tool.setuptools_scm]
diff --git a/tests_golden/integration/test_cookiecutter_generation/api/frequenz-api-test/pyproject.toml b/tests_golden/integration/test_cookiecutter_generation/api/frequenz-api-test/pyproject.toml
index 27cf961..904116b 100644
--- a/tests_golden/integration/test_cookiecutter_generation/api/frequenz-api-test/pyproject.toml
+++ b/tests_golden/integration/test_cookiecutter_generation/api/frequenz-api-test/pyproject.toml
@@ -156,7 +156,7 @@ packages = ["frequenz.api.test"]
 strict = true
 
 [[tool.mypy.overrides]]
-module = ["mkdocs_macros.*", "sybil", "sybil.*"]
+module = ["mkdocs_macros.*"]
 ignore_missing_imports = true
 
 [tool.setuptools.package-dir]
diff --git a/tests_golden/integration/test_cookiecutter_generation/app/frequenz-app-test/pyproject.toml b/tests_golden/integration/test_cookiecutter_generation/app/frequenz-app-test/pyproject.toml
index 3234e24..ae162ed 100644
--- a/tests_golden/integration/test_cookiecutter_generation/app/frequenz-app-test/pyproject.toml
+++ b/tests_golden/integration/test_cookiecutter_generation/app/frequenz-app-test/pyproject.toml
@@ -162,7 +162,7 @@ packages = ["frequenz.app.test"]
 strict = true
 
 [[tool.mypy.overrides]]
-module = ["mkdocs_macros.*", "sybil", "sybil.*"]
+module = ["mkdocs_macros.*"]
 ignore_missing_imports = true
 
 [tool.setuptools_scm]
diff --git a/tests_golden/integration/test_cookiecutter_generation/lib/frequenz-test-python/pyproject.toml b/tests_golden/integration/test_cookiecutter_generation/lib/frequenz-test-python/pyproject.toml
index 1a15745..8ae960f 100644
--- a/tests_golden/integration/test_cookiecutter_generation/lib/frequenz-test-python/pyproject.toml
+++ b/tests_golden/integration/test_cookiecutter_generation/lib/frequenz-test-python/pyproject.toml
@@ -159,7 +159,7 @@ packages = ["frequenz.test"]
 strict = true
 
 [[tool.mypy.overrides]]
-module = ["mkdocs_macros.*", "sybil", "sybil.*"]
+module = ["mkdocs_macros.*"]
 ignore_missing_imports = true
 
 [tool.setuptools_scm]
diff --git a/tests_golden/integration/test_cookiecutter_generation/model/frequenz-model-test/pyproject.toml b/tests_golden/integration/test_cookiecutter_generation/model/frequenz-model-test/pyproject.toml
index e61c4be..c163225 100644
--- a/tests_golden/integration/test_cookiecutter_generation/model/frequenz-model-test/pyproject.toml
+++ b/tests_golden/integration/test_cookiecutter_generation/model/frequenz-model-test/pyproject.toml
@@ -163,7 +163,7 @@ packages = ["frequenz.model.test"]
 strict = true
 
 [[tool.mypy.overrides]]
-module = ["mkdocs_macros.*", "sybil", "sybil.*"]
+module = ["mkdocs_macros.*"]
 ignore_missing_imports = true
 
 [tool.setuptools_scm]
```

llucax commented:
I think we should add a note to the release notes even if it is not a breaking change, as a new feature, like "Add support for pytest 8.x".

Marenz commented:
> Sybil 6 has type information, so we can also remove the mypy exception for it.
> [..]
> it works with only widening the dependencies

But only widening implies we should keep the exception so it still works on the lowest version, no?

llucax commented:
> But only widening implies we should keep the exception so it still works on the lowest version, no?

Yeah, true. I think it would be nice for a smoother transition to first be backwards compatible and we create an issue to remove `mypy`'s exception when we drop support for sybil 5.

llucax committed:
Remove `mypy` exceptions for `sybil`

When running tests, we now use a `sybil` version that is properly typed,
so we don't need to exclude it from `mypy` anymore.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
@Marenz 

* https://github.com/Marenz/frequenz-repo-config-python/pull/2

llucax commented:
I'm thinking that this could be a v0.8.1 release instead, as it is a non breaking change (with the changes proposed in my sub-PR) and it will allow downstream project to upgrade the `pytest` without having to generate the projects files or go through a time-consuming repo-config migration.
